[CSS] Fix :not(:has(...)) invalidation

mdubet opened the pull request:
#### 95ffd81b29e8337bbed14f19e229e5aca0579077
<pre>
[CSS] Fix :not(:has(...)) invalidation
<a href="https://bugs.webkit.org/show_bug.cgi?id=276867">https://bugs.webkit.org/show_bug.cgi?id=276867</a>
<a href="https://rdar.apple.com/122689282">rdar://122689282</a>

Reviewed by Antti Koivisto.

Invalidation of :not(:has(...)) is special because while there can be no
element to actually invalidate, the rule matches.
For ChildChangeInvalidation, we thus propagate the isNegation flag to StyleInvalidator
to invalidate in case of non-matching the :has().
The other *Invalidation classes handle :not() differently, thus the
isNegation flag is purposedly not propagated to StyleInvalidator for them.

* LayoutTests/fast/css/invalidation-not-has-expected.txt: Added.
* LayoutTests/fast/css/invalidation-not-has.html: Added.
* Source/WebCore/style/ChildChangeInvalidation.cpp:
(WebCore::Style::ChildChangeInvalidation::invalidateForChangedElement):
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::evaluateDynamicMediaQueryRules):
* Source/WebCore/style/RuleSet.h:
* Source/WebCore/style/StyleInvalidator.cpp:
(WebCore::Style::Invalidator::Invalidator):
(WebCore::Style::Invalidator::collectRuleInformation):
(WebCore::Style::Invalidator::invalidateIfNeeded):
(WebCore::Style::Invalidator::invalidateUserAgentParts):
(WebCore::Style::Invalidator::addToMatchElementRuleSets):
(WebCore::Style::Invalidator::addToMatchElementRuleSetsRespectingNegation):
(WebCore::Style::Invalidator::invalidateWithScopeBreakingHasPseudoClassRuleSet):
* Source/WebCore/style/StyleInvalidator.h:

Canonical link: <a href="https://commits.webkit.org/285987@main">https://commits.webkit.org/285987@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/db143f92a09e4919035bcb9410270da20e8c33bb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74373 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53802 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27184 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78753 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25611 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76490 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62935 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1587 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58463 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16780 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77440 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48621 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63968 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38866 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45617 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21463 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23944 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67015 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21810 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80272 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1690 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/978 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66742 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1838 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63986 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66026 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16406 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9962 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8118 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1654 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4442 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1683 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1671 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1690 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleInvalidator.cpp:
```diff
@@ -165,11 +172,15 @@ Invalidator::CheckDescendants Invalidator::invalidateIfNeeded(Element& element,
+                break;
+            }
+            if (!matchesAnyRules && ruleSetWithNegation.isNegation == IsNegation::Yes) {
                 element.invalidateStyleInternal();
```
This is not correct and just papers over the problem by invalidating unnecessarily. It will cause massive over-invalidation for :not selectors.

We try invalidation both before and after the state change (ChildChangeInvalidation::invalidateForHasBefore/AfterMutation) and that code should take care of negations.

anttijk commented on LayoutTests/fast/css/invalidation-not-has.html:
```diff
@@ -0,0 +1,40 @@
+<script src="../../resources/testharnessreport.js"></script>
+    <head>
+        <style>
+            div#target:not(:has(> span)) {
+                color: red;
```
would be nice to add WPTs for this (and cover more cases)

anttijk commented on Source/WebCore/style/StyleInvalidator.cpp:
```diff
@@ -166,10 +166,11 @@ Invalidator::CheckDescendants Invalidator::invalidateIfNeeded(Element& element,
 
-            if (ruleCollector.matchesAnyAuthorRules()) {
+            auto matches = ruleCollector.matchesAnyAuthorRules();
+            if ((matches && ruleSet.isNegation == IsNegation::No) || (!matches && ruleSet.isNegation == IsNegation::Yes)) {
```
I think `ruleSet.isNegation == IsNegation::No ? matches : !matches` reads better.

anttijk commented on Source/WebCore/style/ChildChangeInvalidation.cpp:
```diff
@@ -70,7 +70,7 @@ void ChildChangeInvalidation::invalidateForChangedElement(Element& changedElemen
 
         for (auto* selector : invalidationRuleSet.invalidationSelectors) {
-            if (isFirst) {
+            if (isFirst && invalidationRuleSet.isNegation == IsNegation::No) {
```
I think we could maintain this optimization in the negation case by reversing the test. Can be considered separately.

mdubet committed:
[CSS] Fix :not(:has(...)) invalidation
https://bugs.webkit.org/show_bug.cgi?id=276867
rdar://122689282

Reviewed by Antti Koivisto.

Invalidation of :not(:has(...)) is special because while there can be no
element to actually invalidate, the rule matches.
For ChildChangeInvalidation, we thus propagate the isNegation flag to StyleInvalidator
to invalidate in case of non-matching the :has().
The other *Invalidation classes handle :not() differently, thus the
isNegation flag is purposedly not propagated to StyleInvalidator for them.

* LayoutTests/fast/css/invalidation-not-has-expected.txt: Added.
* LayoutTests/fast/css/invalidation-not-has.html: Added.
* Source/WebCore/style/ChildChangeInvalidation.cpp:
(WebCore::Style::ChildChangeInvalidation::invalidateForChangedElement):
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::evaluateDynamicMediaQueryRules):
* Source/WebCore/style/RuleSet.h:
* Source/WebCore/style/StyleInvalidator.cpp:
(WebCore::Style::Invalidator::Invalidator):
(WebCore::Style::Invalidator::collectRuleInformation):
(WebCore::Style::Invalidator::invalidateIfNeeded):
(WebCore::Style::Invalidator::invalidateUserAgentParts):
(WebCore::Style::Invalidator::addToMatchElementRuleSets):
(WebCore::Style::Invalidator::addToMatchElementRuleSetsRespectingNegation):
(WebCore::Style::Invalidator::invalidateWithScopeBreakingHasPseudoClassRuleSet):
* Source/WebCore/style/StyleInvalidator.h:

Canonical link: https://commits.webkit.org/285987@main
